### ISSUE TITLE
Some @Var feedback

### ISSUE THREAD

**Issue by Stephan202** [google/error-prone#484]
I'm investigating converting part of our code base from a "`final` everywhere" approach to a "[`@Var`](http://errorprone.info/bugpattern/Var) when applicable" approach. This looks really promising. I did find some cases where Error-prone does not behave as I had expected. I'll list them here.

1. No warning is emitted for redundant `final` usage in the following constructs:
    * Inside a `try (...)` declaration. (These are [implicitly `final`](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3).)
        ```java
        try (final OutputStream stdin = process.getOutputStream()) {
          // ^ There is no warning for this `stdin` declaration.
        }
        ```
    * Inside a `catch (...)` exception declaration.

        ```java
        try {
          // Some AWS interaction.
        } catch (final AmazonClientException e) {
          // ^ There is no warning for this `e` declaration.
        }
        ```
    * Inside the initializer section of a "regular" three-part `for`-loop:

        ```java
        for (final Iterator<Transformation> it = this.transformations.iterator(); it.hasNext(); ) {
          // ^ There is no warning for this `it` declaration.
        }
        ```
2. No warning is emitted when a non-`@Var` exception declaration inside a `catch (...)`  is redefined.
    ```java
    try {
      // Some AWS interaction.
    } catch (AmazonClientException e) {
      e = null;
      // ^ This is a bad idea in any case (not found in our code base, actually),
      // but Java allows it and there should be a warning that `e` should be annotated `@Var`.
    }
    ```
3. In the initializer section of "regular" three-part `for`-loops `@Var` is not enforced.

    ```java
    for (int i = 0; i < pixels.length; i++) {
      // There's no warning about the fact that `i` is not `@Var`, though it's redefined each iteration.
    }
    ```
    I suspect that this is an explicit design decision. I'd like to argue though that for consistency's sake `for`-loops should not be special-cased. (And observing that in Java 8 many such loops are replaced with stream expressions anyway, I don't think this would hurt users too much.) Alternatively, for this type of loop the guideline could be inverted: disallow `@Var` and require `final` when applicable. (But I'm on the fence about that one, TBH.)

I hope this feedback is useful. Thanks for providing such awesome functionality :)

**Comment by cushon** [comment]
Thanks for the feedback! The issues with try with resources and catch statements are bugs, I'll fix those.

The handling of regular for loops is a [design decision](https://github.com/google/error-prone/blob/390d90ea894e8270f1c9ddc09df20a01feb1647a/core/src/main/java/com/google/errorprone/bugpatterns/VarChecker.java#L63), as you suspected. Initially the check did require `@Var` there, but I got feedback that it was annoying so I special-cased it.
